### ISSUE TITLE
Feature Request: Being able to catch new errors in unmodified files.

### ISSUE THREAD

**Issue by czycha** [golangci/golangci-lint#4226]
### Your feature request related to a problem? Please describe.

Currently when using `--new/--new-from-rev/--new-from-patch`, golangci-lint uses revgrep to only lint the modified lines of code. If needed, a `--whole-files` option can be provided to check the entire file that had been changed.

However, if a change is introduced in one file that would cause an error in another **un**modified file, the error will not be caught.

#### Example

Simple example.

```go
// file: struct.go
package mypackage

// this is targeted by exhaustruct
type MyStruct struct {
  A int
  B string // this is a new line and field!
}
```

```go
// file: client.go
package mypackage

func NewMyStruct() *MyStruct {
  return &MyStruct{
    A: 1,
  }
}
```

Without `--new`,  the error would be caught in the client.go for not initializing with `MyStruct.B`. But with, the `--new` it will not be caught since it was unchanged.

### Describe the solution you'd like.

We need some mechanism for catching such issues.

#### Option: Bisect errors

**Technical flow:**
1. Check errors on the project before the patch is applied.
2. Check errors on the project after the patch is applied.
3. Report which are the net new errors.

**Problems:**
- This doubles the runtime.

**Mitigations:**
- Make this a special flag. Could be `--diff-bisect` or introduce `--diff-mode=bisect|whole-file|change-only`.
- Alternatively, only run this for certain lint rules that we know need context (exhaustruct, deprecations, etc.). Could also use a flag for this.

### Describe alternatives you've considered.

The other option to mitigate for projects would be to just not use the `--new` flags. However, this would require solving any grandfathered lint errors if golangci-lint was incrementally adopted. Additionally, would require updating any of these errors within this vein that have gone uncaught.

**Comment by ldez** [comment]
Hello,

if you want reports from modified/new and unmodified files then the best way is to remove the restriction of the analysis on modified/new files.

I can see several major problems:
- it requires 2 analyses => design problems and performance problems.
- the ratio between complexity (of the feature and the maintenance) and benefit (for the majority of the users) is very low.

We will decline this proposal.

**Comment by czycha** [comment]
Yep makes sense to not implement. Totally understand the complexity involved—just wanted to put this out there as it caught us by surprise on a project that is 2 years old since CI was just not catching those errors due to use of `--new-from-rev` in our CI.

Is it worth at least adding some sort of documentation or warning about this caveat with use of `--new`? Currently, the [description is something like](https://golangci-lint.run/usage/configuration/#command-line-options) "Show only new issues". It should more accurately say something like "Show only issues on modified lines."
